### PR TITLE
Prompt user if a newer version of script is available

### PR DESCRIPTION
**Pull request by harshitbansal05**
@zhangcandrew @henrykironde, the pr that always attempts to install the latest version of scripts.

### CODE REVIEW

**Comment by harshitbansal05** [comment]
@zhangcandrew not sure why the newer builds are failing.

**Comment by henrykironde** [comment]
This Pr has some changes from the previous PR 1358, Let me know if am missing something.

**Comment by harshitbansal05** [comment]
@henrykironde the 2 commits are from the previous pr only. The last commit of this pr adds new changes, which is prompting the user if a newer version of the script is available.

**Comment by henrykironde** [comment]
So do we need to close the other PR?

**Comment by harshitbansal05** [comment]
no @henrykironde, the other pr would download a script if it is not present locally and we need it. This pr would further attempt to always use the latest version of a script. Example, if a script is present locally, this pr would further check upstream if a newer version of that script is available.

**Comment by henrykironde** [comment]
@harshitbansal05 clean this up and we shall be ready to merge

**Comment by harshitbansal05** [comment]
@henrykironde done.

File: retriever/lib/engine_tools.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
 
 from hashlib import md5
 from io import StringIO as NewFile
+from distutils.version import LooseVersion
 from retriever.lib.defaults import HOME_DIR, ENCODING, RETRIEVER_REPOSITORY, RETRIEVER_DATASETS
```
Comment (henrykironde) [inline retriever/lib/engine_tools.py]: remove ENCODING

File: retriever/lib/engine_tools.py [hunk 2]
```diff
@@ -76,11 +77,34 @@ def name_matches(scripts, arg):
...
-
-    from retriever.lib.scripts import get_script_upstream
+            local_version = script.version
+            if arg in RETRIEVER_DATASETS:
+                upstream_version = get_script_version_upstream(arg, repo=RETRIEVER_REPOSITORY)
+            else:
+                upstream_version = get_script_version_upstream(arg)
+            if upstream_version is None or LooseVersion(local_version) >= LooseVersion(upstream_version):
+                return [script]
+            prompt = "A newer version of {dataset} is available. Would you like to download it? (y/N): ".format(dataset=arg)
```
Comment (henrykironde) [inline retriever/lib/engine_tools.py]: line to long

File: retriever/lib/scripts.py [hunk 3]
```diff
@@ -150,59 +176,82 @@ def get_script_upstream(dataset, repo=REPOSITORY):
...
-                r = r.json()
-                for index in range(r['total_count']):
-                    script = r['items'][index]['name']
-                    script = '.'.join(script.split('.')[:-1])
-                    script = script.replace('_', '-')
-                    result_scripts.add(script)
-            except:
-                raise
-    return sorted(result_scripts)
+    try:
```
Comment (zhangcandrew) [inline retriever/lib/scripts.py]: If you want to catch requests exceptions, I think you should catch the base class (requests.exceptions.RequestException). There are a bunch of requests errors that might slip through this check (HTTPError, TimeoutError). 

Now that we are switching over to calls and searches to things upstream, I think this is a good idea to add try catches. I think a better design though would be to abstract retriever specific calls in a new function that solely deals with calls upstream and returning information. This way when something goes wrong with requests, we can report it in retriever without having to try/catch every piece of retriever making calls upstream. The new function could be leveraged by get_dataset_names_upstream and also get_script_version_upstream

**Comment by zhangcandrew** [comment]
It is weird that the checks are failing. Can you make the changes and force push a new commit up? I know ethan was having problems with tests earlier that seemed to clear up so you may have caught some bad timing.

**Comment by henrykironde** [comment]
@harshitbansal05 if you rebase with retriever-recipes-dev the tests will pass.

File: retriever/lib/engine_tools.py (hunk 1 above)
Comment (harshitbansal05) [reply retriever/lib/engine_tools.py]: @henrykironde sorry, but ENCODING is used in the file.

File: retriever/lib/scripts.py (hunk 3 above)
Comment (harshitbansal05) [reply retriever/lib/scripts.py]: @zhangcandrew sorry, your idea is great but could we please do this later in a separate pr?

File: retriever/lib/engine_tools.py [hunk 4]
```diff
@@ -76,11 +77,35 @@ def name_matches(scripts, arg):
...
+                upstream_version = get_script_version_upstream(arg)
+            if upstream_version is None or LooseVersion(local_version) >= LooseVersion(upstream_version):
+                return [script]
+            prompt = "A newer version of {dataset} is available. Would you like to download " \
+                     "it? (y/N): ".format(dataset=arg)
+            should_download = input(prompt)
+            while not (should_download.lower() in ['y', 'n', '']):
+                print("Please enter either y or n.")
+                should_download = input()
+            if should_download.lower() == 'y':
```
Comment (henrykironde) [inline retriever/lib/engine_tools.py]:
```python
            if should_download.lower() == 'y':
                if arg in RETRIEVER_DATASETS:
                    read_script = get_script_upstream(arg, repo=RETRIEVER_REPOSITORY)
                else:
                    read_script = get_script_upstream(arg)
                if read_script is None:
                    print("Unable to download {dataset}.".format(dataset=arg))
                    return [script]
                else:
                    return [read_script]
            else:
                return [script]
```
Change  how return in if statements is done

```python
                if read_script is None:
                    print("Unable to download {dataset}.".format(dataset=arg))
                    return [script]
                return [read_script]
```

**Comment by harshitbansal05** [comment]
@henrykironde done.

File: retriever/lib/scripts.py [hunk 5]
```diff
@@ -150,59 +174,84 @@ def get_script_upstream(dataset, repo=REPOSITORY):
...
+                script = '.'.join(script.split('.')[:-1])
+                script = script.replace('_', '-')
+                scripts.append(script)
+                if len(scripts) == max_scripts:
+                    break
+            return sorted(scripts)
+
+        result_scripts = set()
+        if repo == RETRIEVER_REPOSITORY:
+            search_url = "https://api.github.com/search/code?q={query}+in:file" \
```
Comment (henrykironde) [inline retriever/lib/scripts.py]:
```python
search_base_url = "https://api.github.com/search/code?q={query}+in:file" \
                 "+path:scripts+repo:weecology/"
if repo == RETRIEVER_REPOSITORY:
    search_url = search_base_url + "ret...."
else:
    search_url = search_base_url + "retriever-re.."
```

Comment (harshitbansal05) [reply retriever/lib/scripts.py]: thanks, @henrykironde.

File: retriever/lib/engine_tools.py [hunk 6]
```diff
@@ -76,12 +77,34 @@ def name_matches(scripts, arg):
     if arg == 'all':
         return scripts
 
+    from retriever.lib.scripts import get_script_version_upstream, get_script_upstream
```
Comment (henrykironde) [inline retriever/lib/engine_tools.py]: from retriever.lib.scripts import get_script_version_upstream, get_script_upstream
why do we have the import here?

File: retriever/lib/scripts.py (hunk 3 above)
Comment (zhangcandrew) [reply retriever/lib/scripts.py]: Sure if you're more comfortable with that. I think you can add in the exception change relatively quick, and we can make a note about tackling this in a different PR.

**Comment by henrykironde** [comment]
@zhangcandrew , This looks good but we have a problem with the cyclic dependency and that is why we are putting the imports here https://github.com/weecology/retriever/pull/1362/files#diff-09b72f49b55e7226a62ae92a8a825f82R80 do you have any solution?

**Comment by zhangcandrew** [comment]
I see, I think we can solve this cyclic dependency better if we move the entire name_matches functions into scripts.py. That should stop it. Any thoughts @henrykironde ?

**Comment by henrykironde** [comment]
Let's try that and see

**Comment by harshitbansal05** [comment]
@henrykironde @zhangcandrew I moved the method. Is it fine now?

**Comment by zhangcandrew** [comment]
Fixes the issue and looks good to me! Merging in

**Pull request merged**